// zspex-module/lib/module.js
#!/usr/bin/env node

"use strict";
var fs = require('fs');
var common = require('./common.js');
var mongoose = require('mongoose');
var mod_member = require('./modules/m_member.js');
var mod_team = require('./modules/m_team.js');
var mod_game = require('./modules/m_game.js');
var mod_maps = require('./modules/m_map.js');
var mod_summary = require('./modules/m_summary.js');
var mod_fcm = require('./modules/m_fcm.js');

var configure = require('../../../configure')

var db = require('./util.js');

let ModuleProvider = function () {

    var self = this;

    // modules
    self.Member = mod_member;
    self.Team = mod_team;
    self.Game = mod_game;
    self.Map = mod_maps;

    self.Summary = mod_summary;
    self.Summary.init();
    self.FCM = mod_fcm;

    mongoose.connect(configure.mongodbUri, {
        useMongoClient: true
    });

    var db = mongoose.connection;
    db.on('error', console.error.bind(console, 'connection error:'));
    db.once('open', function callback() {
        console.log('DB connection opened');
    });

    // prototypes
    String.prototype.toRequestID = function () {
        return this.replace("#", "-");
    }

    String.prototype.toDisplayID = function () {
        return this.replace("-", "#");
    }

    String.prototype.toShortID = function () {
        return this.replace(/(#|-).[0-9]+/i, "");
    }

    // Object.prototype.L = function () {
    //     console.dir(this, {
    //         colors: true
    //     });
    // }

}

var handle = (err) => {
    log(err);
    fs.appendFileSync('error.log', err);
}

module.exports = new ModuleProvider();
